optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/service/route/RouteListener.java

```diff
@@ -66,52 +67,73 @@
         bestRoutingPlan = RoutingPlan.empty();
     }
 
+    private static <K, V> List<K> missing(Map<K, Optional<V>> map) {
+        return map.entrySet().stream()
+                .filter(entry -> !entry.getValue().isPresent())
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+    }
+
     @Override
     public void onApplicationEvent(RouteChangedEvent event) {
         // TODO persist the best solution
+
+        // TODO Introduce problem revision (every modification increases revision number, event will only be published
+        //      if revision numbers match).
+
+        // Restore depot
         Location depot = event.depotId().flatMap(locationRepository::find).orElse(null);
-        try {
-            // TODO do this without try-catch
-            Map<Long, Vehicle> vehicleMap = event.vehicleIds().stream()
-                    .collect(Collectors.toMap(vehicleId -> vehicleId, this::findVehicleById));
-            Map<Long, Location> visitMap = event.visitIds().stream()
-                    .collect(Collectors.toMap(visitId -> visitId, this::findLocationById));
-
-            List<RouteWithTrack> routes = event.routes().stream()
-                    // list of deep locations
-                    .map(shallowRoute -> new Route(
-                            vehicleMap.get(shallowRoute.vehicleId),
-                            findLocationById(shallowRoute.depotId),
-                            shallowRoute.visitIds.stream()
-                                    .map(visitMap::get)
-                                    .collect(Collectors.toList())
-                    ))
-                    // add tracks
-                    .map(route -> new RouteWithTrack(route, track(route.depot(), route.visits())))
-                    .collect(Collectors.toList());
-            bestRoutingPlan = new RoutingPlan(
-                    event.distance(),
-                    new ArrayList<>(vehicleMap.values()),
-                    depot,
-                    new ArrayList<>(visitMap.values()),
-                    routes
-            );
-            publisher.publish(bestRoutingPlan);
-        } catch (IllegalStateException e) {
-            logger.warn("Discarding an outdated routing plan: {}", e.toString());
+
+        // Build ID → vehicle map
+        Map<Long, Optional<Vehicle>> vehicleMap = event.vehicleIds().stream()
+                .collect(Collectors.toMap(vehicleId -> vehicleId, this::findVehicleById));
```

Comment:
I *think* I know what you mean by the misuse of `Optional`. By the time I'm collecting the stream results I should already have tested the `Optional` and be using its value, right?

I don't understand how `filter` could help me because I need all of the stream elements at the end. I just want to put non-empty and empty optionals into different buckets.

I thought I could use `partitionBy` to produce a `Map<Boolean, <Map<Long, Vehicle>>` with existing vehicles in the `true` partition and missing vehicles in the `false` partition. But I consider this an overkill:
```java
Map<Boolean, Map<Long, Vehicle>> vehicleMap = event.vehicleIds().stream()
        .map(id -> new Pair<>(id, findVehicleById(id)))
        .collect(Collectors.partitioningBy(
                pair -> pair.getSecond().isPresent(),
                toMap(
                        Pair::getFirst,
                        pair -> pair.getSecond().orElse(null) // NPE
                )
        ));
```
and it will throw an NPE in runtime anyway.

This is an interesting Stream API "puzzle" and I'd like to understand your proposal but please note that this whole approach is probably broken. The goal is to discard new best solutions emitted by solver, that are no longer up-to-date because the user has already made a change (remove vehicle, add visit). We don't want to send such solutions back to the client.

So I was thinking about introducing a revision number, that gets incremented every time a change to the routing problem is made. Any new best solution with a revision number lower than the current revision will be ignored.

Reply:
To be honest, I didn't notice that you're using all of the vehicles at the end. That said, I see that you're trying to solve two problems with one map:

- Find which vehicles are missing.
- Link vehicle IDs to vehicle instances.

I'm assuming you're doing that because `findVehicleById` is prohibitively expensive. Perhaps the solution would be to cache that call? In that case, the problem becomes much smaller and the resulting map is just a `Set<Long>`.